### ISSUE TITLE
`dd.merge` returns incorrect results

### ISSUE THREAD

**Issue by ankravch** [dask/dask#1375]
Before doing smth advanced It would be priceless to make sure that dask does basic operations correctly on largish data - varying npartitions, single machine or cluster compute.

```python
[akravchenko@hnode1 nfs]$ python
Python 3.5.1 |Anaconda custom (64-bit)| (default, Dec  7 2015, 11:16:01)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import dask.dataframe as dd
>>> import pandas as pd
>>> import numpy as np
>>> import dask
>>> from dask.distributed import Executor, progress
>>> from dask.diagnostics import ProgressBar
>>>
>>> f0=pd.DataFrame({'v0':np.random.randint(1,10000000, (100000000,))})
>>> f0_d=dd.from_pandas(f0, npartitions=100)
>>> f1=f0[['v0']].drop_duplicates()
>>> f1_d=dd.from_pandas(f1, npartitions=10) #npartitions=1 seems to give a correct join result
```

where

```python
>>> len(f0)
100000000
>>> len(f1)
9999546
```

join from dask 

```python
>>> res=dd.merge(f0_d,f1_d, on='v0', how='inner').count()
>>> dask.compute(res)
(v0    130892991
dtype: int64,)
>>>
>>> dask.compute(res)
(v0    109156296
dtype: int64,)
>>>
```

and join from pandas:

```python
>>> res_pd=pd.merge(f0, f1, on='v0', how='inner')
>>> res_pd.shape
(100000000, 1)
```

**Comment by jcrist** [comment]
Thanks for the bug report. Just for future reference, if you [wrap code blocks with 3 backticks, they'll format nicely](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code). I added them to your original issue already. 

I don't seem to be able to reproduce your issue. Here's my experience:

```python
In [1]: %paste
>>> import dask.dataframe as dd
>>> import pandas as pd
>>> import numpy as np
>>> import dask

## -- End pasted text --

In [2]: %paste
>>> f0=pd.DataFrame({'v0':np.random.randint(1,10000000, (100000000,))})
>>> f0_d=dd.from_pandas(f0, npartitions=100)
>>> f1=f0[['v0']].drop_duplicates()
>>> f1_d=dd.from_pandas(f1, npartitions=10)

## -- End pasted text --

In [3]: len(f0)
Out[3]: 100000000

In [4]: len(f1)
Out[4]: 9999541

In [5]: res=dd.merge(f0_d,f1_d, on='v0', how='inner')

In [6]: res
Out[6]: dd.DataFrame<hash-jo..., npartitions=100>

In [7]: t = res.compute()   # Do just the join in dask

In [8]: t.count()
Out[8]:
v0    100000000
dtype: int64

In [9]: t2 = res.count().compute()    # Do the join and count in dask

In [10]: t2
Out[10]:
v0    100000000
dtype: int64

In [11]: %paste
>>> res_pd=pd.merge(f0, f1, on='v0', how='inner')
>>> res_pd.shape
(100000000, 1)

## -- End pasted text --

Out[11]: (100000000, 1)
```

I did this a few times, and always got the correct results. It could be a periodic error, or it may be a version issue. What version of dask, pandas, and numpy are you using? If you're not using the latest version of dask (0.10.1) or pandas (0.18.1) can you try again after upgrading?

**Comment by ankravch** [comment]
And from pure Windows:
```python
C:\Users\akravchenko\AppData\Local\Continuum\Anaconda3\Scripts>ipython
Python 3.5.2 |Anaconda 4.1.1 (64-bit)| (default, Jul  5 2016, 11:41:13) [MSC v.1900 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 4.2.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: %paste
>>> import dask.dataframe as dd
>>> import pandas as pd
>>> import numpy as np
>>> import dask

## -- End pasted text --

In [2]: %paste
>>> f0=pd.DataFrame({'v0':np.random.randint(1,10000000, (100000000,))})
>>> f0_d=dd.from_pandas(f0, npartitions=100)
>>> f1=f0[['v0']].drop_duplicates()
>>> f1_d=dd.from_pandas(f1, npartitions=10)

## -- End pasted text --

In [3]: res=dd.merge(f0_d,f1_d, on='v0', how='inner')

In [4]: res
Out[4]: dd.DataFrame<hash-jo..., npartitions=100>

In [5]: t = res.compute()

In [6]: t.count()
Out[6]:
v0    106925853
dtype: int64
```

**Comment by ankravch** [comment]
And on my Mac (OS X 10.10.5):

```python
Antons-MacBook-Pro:~ akravchenko$ ipython
Python 3.5.1 |Anaconda 4.0.0 (x86_64)| (default, Dec  7 2015, 11:24:55) 
Type "copyright", "credits" or "license" for more information.

IPython 4.1.2 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: %paste
>>> import dask.dataframe as dd
>>> import pandas as pd
>>> import numpy as np
>>> import dask

## -- End pasted text --

In [2]: %paste
>>> f0=pd.DataFrame({'v0':np.random.randint(1,10000000, (100000000,))})
>>> f0_d=dd.from_pandas(f0, npartitions=100)
>>> f1=f0[['v0']].drop_duplicates()
>>> f1_d=dd.from_pandas(f1, npartitions=10)

## -- End pasted text --

In [3]: len(f0)
Out[3]: 100000000

In [4]: len(f1)
Out[4]: 9999543

In [5]: res=dd.merge(f0_d,f1_d, on='v0', how='inner')

In [6]: res
Out[6]: dd.DataFrame<hash-jo..., npartitions=100>

In [7]: t = res.compute()

In [8]: t.count()
Out[8]: 
v0    115868592
dtype: int64
```

**Comment by eriknw** [comment]
I duplicated this behavior in Mac OS using dask version 0.10.1 and master in Python 2 and 3 with numpy 1.11.1, pandas 0.18.1.

Reproduced with `conda create -n dask1375 dask ; source activate dask1375` and running this script:
```python
import dask.dataframe as dd
import pandas as pd
import numpy as np
import dask
#from dask.distributed import Executor, progress
from dask.diagnostics import ProgressBar

N = 100000000
f0=pd.DataFrame({'v0':np.random.randint(1,N//10, (N,))})
f0_d=dd.from_pandas(f0, npartitions=100)
f1=f0[['v0']].drop_duplicates()
f1_d=dd.from_pandas(f1, npartitions=10)
print(len(f0))
print(len(f1))
res=dd.merge(f0_d,f1_d, on='v0', how='inner').count()
print(dask.compute(res))
print(dask.compute(res))
res_pd=pd.merge(f0, f1, on='v0', how='inner')
print(res_pd.shape)
```

**Comment by mrocklin** [comment]
I'll try to take a look at this today

**Comment by mrocklin** [comment]
I believe that this is resolved in #1383

**Comment by mrocklin** [comment]
This likely also solves http://stackoverflow.com/questions/38416836/result-of-join-in-dask-dataframes-seems-to-depend-on-the-way-the-dask-datafram

**Comment by ankravch** [comment]
Hi Matt, is a fix for #1383 already merged in dask?

**Comment by mrocklin** [comment]
Not yet.  I'm rolling a few fixes in there.  I wouldn't be surprised if it was merged late tomorrow.
